activerecord/lib/active_record/dynamic_matchers.rb

```diff
@@ -2,121 +2,139 @@
 
 module ActiveRecord
   module DynamicMatchers #:nodoc:
-    private
-      def respond_to_missing?(name, _)
-        if self == Base
-          super
-        else
-          match = Method.match(self, name)
-          match && match.valid? || super
-        end
+    def self.included(base)
+      base.extend(ClassMethods)
+    end
+
+    module ClassMethods
+      def inherited(child_class)
+        child_class.generated_dynamic_matchers
+        super
       end
 
-      def method_missing(name, *arguments, &block)
-        match = Method.match(self, name)
-
-        if match && match.valid?
-          match.define
-          send(name, *arguments, &block)
-        else
-          super
+      def generated_dynamic_matchers
+        @generated_dynamic_matchers ||= begin
+          mod = Module.new
+          extend mod
+          mod
         end
       end
 
-      class Method
-        @matchers = []
-
-        class << self
-          attr_reader :matchers
-
-          def match(model, name)
-            klass = matchers.find { |k| k.pattern.match?(name) }
-            klass.new(model, name) if klass
-          end
-
-          def pattern
-            @pattern ||= /\A#{prefix}_([_a-zA-Z]\w*)#{suffix}\Z/
+      private
+        def respond_to_missing?(name, _)
+          if self == Base
+            super
+          else
+            match = Method.match(self, name)
```

Comment:
Since we're changing this anyway, how about: `Method.match(self, name)&.valid? || super`?